typeinfo/typeinfo.go

```diff
@@ -0,0 +1,137 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+var cmutex sync.RWMutex
+var cache = make(map[string]Info)
+
+// Return the Info of a type name passed as parameter
+// Returns error if the type has not been reflected yet.
+func GetInfoFromName(typeName string) (Info, error) {
```

Comment:
This function is for use in a future `prepare()` function that is already [prototyped](https://github.com/Aflynn50/sqlair/blob/prepare/internal/prepare/prepare.go#L22). When preparing the query, at some point we have the textual name of the type (as in `Person` if the query contains `SELECT col_foo AS &Person.Foo ...`)

Reply:
To address your point about two types having the same name:

In the prepare prototype the flow is:
	- generate reflect information for all the (empty instantiations of) types and store it in the cache
	- cycle through the IO expressions in the parsed query and for each Go object specified (`&Person.name`), look up the reflection information using GetInfoFromName
	- Use this reflection info to run relevant checks (such as the tag written in the IO expression being a member of the Go struct, or generating the list of columns for an asterisk expression

In the [specification](https://docs.google.com/document/d/1FbRZisjK0z7shy8m1cHzzEMDpxGwsK-54xDpf9HY4Xo/edit#) we have the condition that two types should not have the same name. There should be a type alias declared for instances where the user wants to scan two separate objects of the same type. This code was designed with that condition in mind.

It is also worth noting that these cached objects are only being used for their type information; not their actual value. So for instances such as certain types of Join where the user might write the same object in the query twice, they still only have to pass the a single object to prepare. Also the order in which they pass the objects does not matter, it can be any order.

Reply:
`s/GetInfoForName/InfoForName/`

Reply:
> In specification we have the condition that two types should not have the same name.

The API we designed in the specification is able to handle such cases properly. The constraint for type naming is per expression, not globally, otherwise imagine what happens if we have 5 different `DataInfo` objects, created by 5 different packages written by 5 different authors, in the overall application. How would SQLair pick up the right data for each type? Insanity, right?

Reply:
Ah ok that makes sense. I had a different idea in my head of how the cache would be used (I was thinking it would be limited to a single query for caching info from prepare to use in scan, rather than global). For this use case a local map in prepare would make more sense.

Reply:
Nice, makes sense.

Reply:
No `InfoFromName` function and the cache is back to use `reflect.Type` as key.